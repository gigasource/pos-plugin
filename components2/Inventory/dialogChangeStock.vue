<script>
export default {
  setup() {
    return () => <>
      <g-dialog v-model={internalValue}>
        <div class="dialog">
          <div class="dialog-title">
            {t('inventory.addRemoveStock')} </div>
          <g-icon size="16" class="dialog-icon--close" onClick={() => internalValue = false}>
            icon-close
          </g-icon>
          <div class="dialog-content">
            <p>
              <b>
                {t('inventory.item')}: </b>
              {name} </p>
            <p>
              <b>
                {t('inventory.currentStock')}: </b>
              {$filters.formatCurrency(stock)} </p>
            <p>
              <b>
                {t('inventory.newStock')}: </b>
              <span style={{ ...mode === 'add' && { color: '#1271FF' }, ...mode === 'remove' && { color: '#FF4452' } }}>
                {$filters.formatCurrency(newStock)} </span>
            </p>
            <div class="dialog-content__action">
              <div class={['btn', mode === 'add' && 'btn--blue']} onClick={() => mode = 'add'}>
                <g-icon color="white">
                  add
                </g-icon>
              </div>
              <div class={['btn', mode === 'remove' && 'btn--red', !removeable && 'disabled']} onClick={() => mode = 'remove'}>
                <g-icon color="white">
                  remove
                </g-icon>
              </div>
              <g-text-field-bs ref={textfield} rules={rules} modelValue={change} onUpdate:modelValue={changeValue}>
              </g-text-field-bs>
            </div>
            {
              (mode === 'remove') ?
                  <g-select class="dialog-content__reason" text-field-component="GTextFieldBs" items={reasons} v-model={reason} placeholder="Reason (Optional)">
                  </g-select>
                  :

                  <div class="dialog-content__reason" style="height: 38px"></div>
            }
          </div>
          <div class="dialog-keyboard">
            <pos-keyboard-full width="100%" type="numeric" onEnterPressed={submit}>
            </pos-keyboard-full>
          </div>
        </div>
      </g-dialog>
    </>
  }
}
</script>

<style scoped lang="scss">
.dialog {
  width: 500px;
  margin-left: auto;
  margin-right: auto;
  background-color: white;
  padding: 16px;
  position: relative;
  overflow: auto;
  border-radius: 2px;

  &-title {
    font-size: 20px;
    line-height: 25px;
    font-weight: 700;
    margin-bottom: 8px;
  }

  &-icon--close {
    position: absolute;
    top: 16px;
    right: 16px;
  }

  &-content {
    p {
      font-size: 16px;
      line-height: 20px;
      margin: 8px 0;
    }

    .btn {
      flex: 0 0 38px;
      height: 38px;
      display: flex;
      align-items: center;
      justify-content: center;
      background-color: #9e9e9e;
      border-radius: 4px;
      margin-right: 8px;

      &--blue {
        background-color: #1271FF;
      }

      &--red {
        background-color: #ff4552;
      }
    }

    .bs-tf-wrapper,
    .g-select ::v-deep .bs-tf-wrapper {
      margin: 0;
      width: 100%;

      ::v-deep input {
        outline: none;
      }
    }

    &__action {
      display: flex;
      align-items: center;
      margin-bottom: 8px;
    }

    &__reason {
      margin-bottom: 8px;
    }
  }

  &-keyboard {
    background: #bdbdbd;
    padding: 4px;
    border-radius: 2px;
  }
}

@media screen and (max-height: 599px) {
  .dialog {
    width: 400px;
    padding: 12px;

    &-title {
      font-size: 18px;
      line-height: 22px;
    }

    &-content {

      p {
        font-size: 14px;
        line-height: 18px;
        margin: 4px 0;
      }

      &__action {
        margin-bottom: 4px;
      }

      &__reason {
        margin-bottom: 4px;
      }
    }

    &-keyboard {
      ::v-deep .key {
        font-size: 16px !important;
      }
    }
  }
}
</style>
