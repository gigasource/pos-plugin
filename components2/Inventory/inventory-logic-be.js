import cms from 'cms'
import {
	inventories,
	inventoryCategories,
	inventoryHistories
} from './inventory-logic-ui'
import {ObjectID} from "bson"
import _ from 'lodash'

const Inventory = cms.getModel('Inventory')
const InventoryCategory = cms.getModel('InventoryCategory')
const InventoryHistory = cms.getModel('InventoryHistory')

export async function loadInventories() {
	const _inventories = await Inventory.find()
	inventories.value = _inventories.map(item => ({
		...item,
		category: inventoryCategories.value.find(cate => cate._id.toString() === item.category.toString())
	}))
}

export async function loadInventoryCategories() {
	inventoryCategories.value = await InventoryCategory.find()
}

export async function loadInventoryHistories() {
	inventoryHistories.value = await InventoryHistory.find()
}

export async function createInventory(inventory) {
	if (!inventory.id) {
		const maxId = inventories.value.reduce((maxId, inventory) => {
			return Math.max(maxId, parseInt(inventory.id))
		}, 0)
		inventory.id = (maxId ? maxId + 1 : 0)
	} else {
		const foundId = inventories.value.find(_inventory => {
			return inventory.id === _inventory.id
		})
		if (foundId) return
	}
	inventory.lastUpdateTimestamp = new Date()
	inventories.value.push(_.cloneDeep(inventory))
	inventory.category = (typeof inventory.category === 'Object' ? inventory.category._id : inventory.category)
	await Inventory.create(inventory)
}

export async function updateInventory({ _id, name, category, unit, stock, lowStockThreshold }, reason) {
	const inventory = _.find(inventories.value, (inventory) => inventory._id.toString() === _id.toString())
	Object.assign(inventory, {
		name,
		category,
		unit,
		stock,
		lowStockThreshold,
		lastUpdateTimestamp: new Date()
	})
	category = (typeof category === 'object' ? category._id : category)
	if (stock !== inventory.stock) {
		const history = {
			inventory: _id,
			category,
			type: stock.value > inventory.stock ? 'add' : 'remove',
			amount: Math.abs(stock.value - inventory.stock),
			date: new Date(),
			reason: reason ? reason : 'Update stock'
		}
		await updateInventoryHistory(history)
	}
	await Inventory.findOneAndUpdate(
		{ _id },
		{
			name, category, unit, stock, lowStockThreshold, lastUpdateTimestamp: new Date()
		}
	)
}

export async function updateInventoryHistory(history) {
	inventoryHistories.value.push(history)
	await InventoryHistory.create(history)
}

export async function deleteInventory(ids) {
	_.remove(inventories.value, (inventory) => {
		return ids.includes(inventory._id)
	})
	const inventoriesDeleted = await Inventory.find({_id: {$in: ids}})
	const result = await Inventory.deleteMany({_id: {$in: ids}})
	if(result.n === ids.length) {
		const histories = inventoriesDeleted.value.map(i => ({
			inventory: i._id.toString(),
			category: i.category,
			type: 'remove',
			amount: i.stock,
			date: new Date(),
			reason: 'Remove inventory',
		}))
		await updateInventoryHistory(histories)
	}
}
