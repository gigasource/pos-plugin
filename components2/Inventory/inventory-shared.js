import { ref } from 'vue'
import cms from 'cms'
import { ObjectID } from 'bson'

const INVENTORY_COL = 'Inventory'
const INVENTORY_CATEGORY_COL = 'InventoryCategory'
const INVENTORY_UNIT_COL = 'InventoryUnit'
const INVENTORY_HISTORY_COL = 'InventoryHistory'
const COMMIT_COL = 'OrderCommit'

const inventories = ref([])
const inventoryCategories = ref([])
const inventoriesLoadTimestamp = ref(null)

export async function updateInventoryHistory(history) {
  await cms.getModel(INVENTORY_HISTORY_COL).create(history)
}

export async function loadInventories(filters = []) {
  const condition = filters.reduce((acc, filter) => ({...acc, ...filter['condition']}), {})
  const _inventories = await cms.getModel(INVENTORY_COL).find(condition)
  const _categories = await cms.getModel(INVENTORY_CATEGORY_COL).find()
  inventories.value = _inventories.map(item => ({
    ...item,
    category: _categories.find(cate => cate._id.toString() === item.category)
  }))
  inventoriesLoadTimestamp.value = new Date().getTime()
}

export async function loadInventoryCategories() {
  inventoryCategories.value = await cms.getModel(INVENTORY_CATEGORY_COL).find({})
}

export async function updateInventory({ _id, name, category, unit, stock, lowStockThreshold }) {
  await cms.getModel(INVENTORY_COL).findOneAndUpdate(
    { _id: new ObjectID() },
    {
      name, category, unit, stock, lowStockThreshold, lastUpdateTimestamp: new Date()
    }
  )
}

export async function  deleteInventory(ids) {
  ids = ids.map(id => ObjectID(id))
  const inventories = await cms.getModel(INVENTORY_COL).find({_id: {$in: ids}})
  const result = await cms.getModel(INVENTORY_COL).deleteMany({_id: {$in: ids}})
  if(result.n === ids.length) {
    const histories = inventories.value.map(i => ({
      inventory: i._id.toString(),
      category: i.category,
      type: 'remove',
      amount: i.stock,
      date: new Date(),
      reason: 'Remove inventory',
    }))
    await updateInventoryHistory(histories)
  }
}
