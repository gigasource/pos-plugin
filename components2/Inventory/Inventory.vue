<script>
import { onBeforeMount } from 'vue'
import { useI18n } from 'vue-i18n'
import PosRangeSlider from '../pos-shared-components/POSInput/PosRangeSlider'
import PosTextfieldNew from '../pos-shared-components/POSInput/PosTextfieldNew'
import dialogFormInput from '../pos-shared-components/dialogFormInput'
import dialogChangeStock from './helpers/dialogChangeStock'
import dialogInventoryCategory from './helpers/dialogInventoryCategory'
import dialogInventoryRetailCategory from './helpers/dialogInventoryRetailCategory'

import {
  renderInventoryDialog,
  renderChangeStockDialog,
  renderCategoryDialog,
  renderFilterDialog
} from './inventory-ui-shared'


import {
  loadInventories
} from './inventory-logic-be'

import {
  loadProducts,
  loadCategories
} from '../Product/product-logic-be';

import {
  renderMainInventoryTable
} from './inventory-table'
import { genScopeId } from '../utils';

export default {
  name: 'Inventory',
  components: {PosRangeSlider, PosTextfieldNew, dialogFormInput, dialogChangeStock, dialogInventoryCategory, dialogInventoryRetailCategory},
  setup(props, { emit }) {
    const { t } = useI18n();

    onBeforeMount(async () => {
      await loadCategories()
      await loadProducts()
      await loadInventories()
    })

    const {
      renderMainTable,
      renderToolBar
    } = renderMainInventoryTable(props, { emit })

    return genScopeId(() => <>
      <div style="height: 100%; display: flex; flex-direction: column">
        {renderMainTable()}
        {renderToolBar()}
        {renderInventoryDialog(t)}
        {renderChangeStockDialog()}
        {renderCategoryDialog()}
        {renderFilterDialog(t)}
      </div>
    </>)
  }
}
</script>
<style scoped lang="scss">
.g-table {
  height: calc(100% - 64px);

  ::v-deep table {
    table-layout: fixed;
  }

  thead tr th {
    font-size: 13px;
    color: #1d1d26;
    padding: 0 8px;
    background-color: #fff;
    cursor: pointer;
    text-align: left;
    -webkit-tap-highlight-color: transparent;
  }

  tr td {
    padding: 0 8px;
    font-size: 13px;
    line-height: 16px;
    height: 33px
  }

  tr td:first-child,
  tr th:first-child {
    padding-right: 0;
  }

  tr {
    td:nth-child(1),
    th:nth-child(1) {
      width: 5%;
    }

    td:nth-child(2):not(.filter-wrapper),
    th:nth-child(2) {
      width: 10%;
    }

    td:nth-child(3),
    th:nth-child(3) {
      width: 30%;
    }

    td:nth-child(4),
    th:nth-child(4) {
      width: 15%;
    }

    td:nth-child(5),
    th:nth-child(5) {
      width: 15%;
    }

    td:nth-child(6),
    th:nth-child(6) {
      width: 10%;
    }

    td:nth-child(7),
    th:nth-child(7) {
      width: 10%;
    }
  }

  .sticky {
    td {
      position: sticky;
      z-index: 2;
      top: 48px;
    }
  }

  .filter-wrapper {
    background-color: #bdbdbd;
    height: 48px;

    .filter {
      color: #1d1d26;
      font-size: 13px;
      line-height: 16px;
      font-weight: 700;
      display: flex;
      align-items: center;

      .group-chip {
        display: flex;
        flex-wrap: nowrap;
        overflow: auto;
        margin: 0 4px;

        &::-webkit-scrollbar {
          display: none;
        }

        ::v-deep .g-chip {
          overflow: visible;
        }

        .chip-title {
          color: #97A3B4;
          font-weight: 400;
          font-size: 11px;
        }

        .chip-content {
          color: #1D1D26;
          font-weight: 700;
          font-size: 12px;
        }
      }

      .btn-add-filter {
        border-radius: 4px;
        background-color: #2979ff;
        color: white;
        padding: 10px;
        cursor: pointer;
        font-size: 14px;
      }
    }
  }
}

.g-select ::v-deep .g-tf-wrapper {
  margin: 16px 0 4px;
  width: 100%;

  .g-tf-input--fake-caret span {
    color: rgba(0, 0, 0, 0.87)
  }

  fieldset {
    border-color: #C9C9C9;
    border-radius: 2px;
  }

  &.g-tf__focused fieldset {
    border-color: #1867c0;
  }

  .g-tf-label {
    font-weight: bold;
    color: #1D1D26;
  }

  .input {
    padding-left: 6px;
  }
}

@media screen and (max-height: 599px) {
  .g-table {
    thead tr th {
      height: 36px;
    }

    tr td {
      height: 28px;
    }

    .filter-wrapper {
      height: 36px;

      .filter {
        .btn-add-filter {
          padding: 4px 8px;
          font-size: 12px;
        }
      }
    }
  }
}

@media screen and (max-width: 1023px) {
  .g-toolbar {
    .g-btn ::v-deep .g-btn__content {
      font-size: 0;

      .g-icon {
        margin-right: 0 !important;
        font-size: 24px !important;
        width: 24px !important;
        height: 24px !important;
      }
    }
  }
}
</style>

<style lang="scss">
@media screen and (max-height: 599px) {
  .menu-select-inventory {
    .g-list {
      .g-list-item {
        min-height: 0;
        padding: 4px;

        .g-list-item-content {
          margin: 0;

          .g-list-item-text {
            text-align: left;
          }
        }
      }
    }
  }
}
</style>
