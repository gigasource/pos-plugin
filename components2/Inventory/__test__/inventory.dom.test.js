jest.setTimeout(60000)

import { makeWrapper, orm, wrapper } from '../../test-utils';
import {nextTick} from "vue";
import Inventory from '../Inventory'
import {
  dialog
} from "../inventory-ui-shared";
import _ from 'lodash'
import {
  inventories,
  inventoryCategories,
  filter
} from "../inventory-logic-ui";
import {
  addedCategory
} from "../helpers/dialog-inventory-category-logic";

const { prepareInventoryDb } = require('../../../backend/inventory/inventory.prepare.test')

const delay = require("delay");

jest.mock('cms', () => {
  const { cmsFactory } = require('../../../test-utils')
  const _cms = cmsFactory('inventoryDom')
  global.cms = _cms
  return {
    socket: _cms.feSocket,
    getModel: function (modelName) {
      return _cms.orm(modelName)
    }
  }
})

let cms = global.cms

const {stringify} = require("schemahandler/utils");

describe('Test inventory screen', function () {
  beforeAll(async () => {
    await cms.initDemoData()
    await prepareInventoryDb(cms.orm)
    cms.triggerFeConnect()
  })

  it('Case 1: Inventory', async () => {
    makeWrapper(Inventory, {
      shallow: true,
      props: {
      }
    });
    await nextTick();
    await delay(50);
    expect(wrapper.html()).toMatchSnapshot()
    expect(inventories.value).toMatchSnapshot()
    expect(inventoryCategories.value).toMatchSnapshot()
  })

  it('Case 2: Category dialog', async () => {
    dialog.value.category = true
    makeWrapper(Inventory, {
      shallow: false,
      props: {
      }
    });
    await nextTick();
    await delay(50);
    const component = wrapper.find("[content-class='dialog-inventory-category']")
    /**
     * First render
     */
    expect(component.html()).toMatchSnapshot()
    /**
     * Add 1 category
     */
    const addCategoryBtn = component.find('[data-jest-addCategory]')
    expect(addCategoryBtn.exists()).toBe(true)
    const oldLength = component.findAll("[class='category-item']").length
    addCategoryBtn.trigger('click')
    await nextTick()
    const categoryItems = wrapper.findAll("[class='category-item']")
    expect(categoryItems.length).toEqual(oldLength + 1)
    const _addedCategory = _.last(categoryItems)
    const textField = _addedCategory.find('g-text-field-bs-stub')
    await textField.trigger('click')
    /**
     * Fake set value
     */
    const a = addedCategory.value
    addedCategory.value[0].name = 'TestCategory'
    await nextTick()
    const completeCategoryBtn = wrapper.find('[data-jest-complete]')
    await completeCategoryBtn.trigger('click')
    await nextTick()
    /**
     * Expect re-render of dialog after open keyboard
     */
    expect(wrapper.find("[content-class='dialog-inventory-category']").html()).toMatchSnapshot()
    await delay(50)
    expect(inventoryCategories.value).toMatchSnapshot()
  })

  it('Case 3: Filter dialog', async () => {
    dialog.value.filter = true
    makeWrapper(Inventory, {
      shallow: false,
      props: {
      }
    });
    await nextTick();
    await delay(50);
    const component = wrapper.find("[data-jest-filter-dialog]")
    expect(component.exists()).toBe(true)
    /**
     * Fake filter action
     */
    filter.value = {
      id: 1,
      stock: [10, 20]
    }
    await nextTick()
    const showFilter = wrapper.find("[class='filter-wrapper']")
    expect(showFilter.html()).toMatchSnapshot()
  })
})

