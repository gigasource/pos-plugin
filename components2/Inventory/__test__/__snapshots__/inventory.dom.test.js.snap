// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test inventory screen Case 1: Inventory 1`] = `
<div style="height: 100%; display: flex; flex-direction: column;">
  <g-simple-table-stub striped="true"
                       fixed-header="true"
                       style="flex: 1;"
  >
    <thead>
      <tr>
        <th>
        </th>
        <th>
          ID
        </th>
        <th>
          article.name
          <g-icon-stub size="12">
            mdi-filter
          </g-icon-stub>
        </th>
        <th>
          inventory.lastUpdate
        </th>
        <th>
          article.category
          <g-icon-stub size="12">
            mdi-magnify
          </g-icon-stub>
        </th>
        <th>
          inventory.unit
        </th>
        <th>
          inventory.stock
        </th>
      </tr>
    </thead>
    <tr>
      <td class="bg-grey-lighten-1">
        <g-checkbox-stub modelvalue
                         value="0000000000000000000007d2,0000000000000000000007d4,0000000000000000000007d6,0000000000000000000007d8,0000000000000000000007da,0000000000000000000007dc"
                         multiple="true"
        >
        </g-checkbox-stub>
      </td>
      <td colspan="6"
          class="filter-wrapper"
      >
        <div class="filter">
          settings.filter
          <div class="group-chip">
          </div>
          <g-spacer-stub>
          </g-spacer-stub>
          <div class="btn-add-filter">
            + inventory.addFilter
          </div>
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <g-checkbox-stub modelvalue
                         value="0000000000000000000007d2"
        >
        </g-checkbox-stub>
      </td>
      <td>
        1
      </td>
      <td>
        Fish
      </td>
      <td>
        05/01/2021 00:00
      </td>
      <td>
        Food
      </td>
      <td>
        kg
      </td>
      <td>
        <div class="row-flex justify-between">
          30
          <g-icon-stub size="18"
                       color="#757575"
          >
            edit
          </g-icon-stub>
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <g-checkbox-stub modelvalue
                         value="0000000000000000000007d4"
        >
        </g-checkbox-stub>
      </td>
      <td>
        2
      </td>
      <td>
        Meat
      </td>
      <td>
        05/01/2021 00:00
      </td>
      <td>
        Food
      </td>
      <td>
        kg
      </td>
      <td>
        <div class="row-flex justify-between">
          50
          <g-icon-stub size="18"
                       color="#757575"
          >
            edit
          </g-icon-stub>
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <g-checkbox-stub modelvalue
                         value="0000000000000000000007d6"
        >
        </g-checkbox-stub>
      </td>
      <td>
        3
      </td>
      <td>
        Pork
      </td>
      <td>
        05/01/2021 00:00
      </td>
      <td>
        Food
      </td>
      <td>
        kg
      </td>
      <td>
        <div class="row-flex justify-between">
          60.22
          <g-icon-stub size="18"
                       color="#757575"
          >
            edit
          </g-icon-stub>
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <g-checkbox-stub modelvalue
                         value="0000000000000000000007d8"
        >
        </g-checkbox-stub>
      </td>
      <td>
        4
      </td>
      <td>
        Apple
      </td>
      <td>
        05/01/2021 00:00
      </td>
      <td>
        Fruit
      </td>
      <td>
        piece
      </td>
      <td>
        <div class="row-flex justify-between">
          100
          <g-icon-stub size="18"
                       color="#757575"
          >
            edit
          </g-icon-stub>
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <g-checkbox-stub modelvalue
                         value="0000000000000000000007da"
        >
        </g-checkbox-stub>
      </td>
      <td>
        5
      </td>
      <td>
        Banana
      </td>
      <td>
        05/01/2021 00:00
      </td>
      <td>
        Fruit
      </td>
      <td>
        piece
      </td>
      <td>
        <div class="row-flex justify-between">
          10
          <g-icon-stub size="18"
                       color="#757575"
          >
            edit
          </g-icon-stub>
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <g-checkbox-stub modelvalue
                         value="0000000000000000000007dc"
        >
        </g-checkbox-stub>
      </td>
      <td>
        6
      </td>
      <td>
        Vodka
      </td>
      <td>
        05/01/2021 00:00
      </td>
      <td>
        Drink
      </td>
      <td>
        l
      </td>
      <td>
        <div class="row-flex justify-between">
          5.50
          <g-icon-stub size="18"
                       color="#757575"
          >
            edit
          </g-icon-stub>
        </div>
      </td>
    </tr>
  </g-simple-table-stub>
  <div>
    <g-toolbar-stub color="#eeeeee"
                    elevation="0"
    >
      <g-btn-stub uppercase="false"
                  style="margin-right: 5px;"
      >
        <g-icon-stub small="true"
                     style="margin-right: 5px;"
        >
          icon-back
        </g-icon-stub>
        ui.back
      </g-btn-stub>
      <g-btn-stub uppercase="false">
        <g-icon-stub small="true"
                     style="margin-right: 5px;"
        >
          icon-inventory-report
        </g-icon-stub>
        inventory.report
      </g-btn-stub>
      <g-spacer-stub>
      </g-spacer-stub>
      <g-btn-stub uppercase="false"
                  style="margin-right: 5px;"
      >
        <g-icon-stub small="true"
                     style="margin-right: 5px;"
        >
          icon-inventory-new-stock
        </g-icon-stub>
        inventory.newStock
      </g-btn-stub>
      <g-btn-stub uppercase="false"
                  style="margin-right: 5px;"
      >
        <g-icon-stub small="true"
                     style="margin-right: 5px;"
        >
          icon-inventory-category
        </g-icon-stub>
        article.category
      </g-btn-stub>
      <g-btn-stub disabled="true"
                  uppercase="false"
                  style="margin-right: 5px;"
      >
        <g-icon-stub small="true"
                     style="margin-right: 5px;"
        >
          icon-inventory-delete
        </g-icon-stub>
        ui.delete
      </g-btn-stub>
      <g-btn-stub disabled="true"
                  uppercase="false"
                  style="margin-right: 5px;"
      >
        <g-icon-stub small="true"
                     style="margin-right: 5px;"
        >
          icon-inventory-edit
        </g-icon-stub>
        ui.edit
      </g-btn-stub>
      <g-btn-stub uppercase="false"
                  background-color="#4CAF50"
                  text-color="#FFF"
      >
        <span style="font-size: 14px !important;">
          + inventory.newProduct
        </span>
      </g-btn-stub>
    </g-toolbar-stub>
  </div>
  <dialog-form-input-stub>
  </dialog-form-input-stub>
  <dialog-change-stock-stub>
  </dialog-change-stock-stub>
  <dialog-inventory-category-stub>
  </dialog-inventory-category-stub>
  <dialog-form-input-stub>
  </dialog-form-input-stub>
</div>
`;

exports[`Test inventory screen Case 1: Inventory 2`] = `
Array [
  Object {
    "_id": "0000000000000000000007d2",
    "category": Object {
      "_id": "0000000000000000000007cf",
      "available": false,
      "name": "Food",
    },
    "id": "1",
    "lastUpdateTimestamp": 2021-01-04T17:00:00.000Z,
    "name": "Fish",
    "stock": 30,
    "unit": "kg",
  },
  Object {
    "_id": "0000000000000000000007d4",
    "category": Object {
      "_id": "0000000000000000000007cf",
      "available": false,
      "name": "Food",
    },
    "id": "2",
    "lastUpdateTimestamp": 2021-01-04T17:00:00.000Z,
    "name": "Meat",
    "stock": 50,
    "unit": "kg",
  },
  Object {
    "_id": "0000000000000000000007d6",
    "category": Object {
      "_id": "0000000000000000000007cf",
      "available": false,
      "name": "Food",
    },
    "id": "3",
    "lastUpdateTimestamp": 2021-01-04T17:00:00.000Z,
    "name": "Pork",
    "stock": 60.22,
    "unit": "kg",
  },
  Object {
    "_id": "0000000000000000000007d8",
    "category": Object {
      "_id": "0000000000000000000007d0",
      "available": false,
      "name": "Fruit",
    },
    "id": "4",
    "lastUpdateTimestamp": 2021-01-04T17:00:00.000Z,
    "name": "Apple",
    "stock": 100,
    "unit": "piece",
  },
  Object {
    "_id": "0000000000000000000007da",
    "category": Object {
      "_id": "0000000000000000000007d0",
      "available": false,
      "name": "Fruit",
    },
    "id": "5",
    "lastUpdateTimestamp": 2021-01-04T17:00:00.000Z,
    "name": "Banana",
    "stock": 10,
    "unit": "piece",
  },
  Object {
    "_id": "0000000000000000000007dc",
    "category": Object {
      "_id": "0000000000000000000007d1",
      "available": false,
      "name": "Drink",
    },
    "id": "6",
    "lastUpdateTimestamp": 2021-01-04T17:00:00.000Z,
    "name": "Vodka",
    "stock": 5.5,
    "unit": "l",
  },
]
`;

exports[`Test inventory screen Case 1: Inventory 3`] = `
Array [
  Object {
    "_id": "0000000000000000000007d3",
    "amount": 30,
    "category": "0000000000000000000007cf",
    "date": 2021-01-04T17:00:00.000Z,
    "inventory": "0000000000000000000007d2",
    "type": "add",
  },
  Object {
    "_id": "0000000000000000000007d5",
    "amount": 50,
    "category": "0000000000000000000007cf",
    "date": 2021-01-04T17:00:00.000Z,
    "inventory": "0000000000000000000007d4",
    "type": "add",
  },
  Object {
    "_id": "0000000000000000000007d7",
    "amount": 60.22,
    "category": "0000000000000000000007cf",
    "date": 2021-01-04T17:00:00.000Z,
    "inventory": "0000000000000000000007d6",
    "type": "add",
  },
  Object {
    "_id": "0000000000000000000007d9",
    "amount": 100,
    "category": "0000000000000000000007d0",
    "date": 2021-01-04T17:00:00.000Z,
    "inventory": "0000000000000000000007d8",
    "type": "add",
  },
  Object {
    "_id": "0000000000000000000007db",
    "amount": 10,
    "category": "0000000000000000000007d0",
    "date": 2021-01-04T17:00:00.000Z,
    "inventory": "0000000000000000000007da",
    "type": "add",
  },
  Object {
    "_id": "0000000000000000000007dd",
    "amount": 5.5,
    "category": "0000000000000000000007d1",
    "date": 2021-01-04T17:00:00.000Z,
    "inventory": "0000000000000000000007dc",
    "type": "add",
  },
]
`;

exports[`Test inventory screen Case 1: Inventory 4`] = `
Array [
  Object {
    "_id": "0000000000000000000007cf",
    "available": false,
    "name": "Food",
  },
  Object {
    "_id": "0000000000000000000007d0",
    "available": false,
    "name": "Fruit",
  },
  Object {
    "_id": "0000000000000000000007d1",
    "available": false,
    "name": "Drink",
  },
]
`;

exports[`Test inventory screen Case 2: Category dialog 1`] = `
<g-dialog-stub fullscreen="true"
               modelvalue="true"
               content-class="dialog-inventory-category"
>
  <div class="dialog">
    <div class="dialog-center">
      <div class="category">
        <div class="category-item">
          <g-text-field-bs-stub rules="val => _inventoryLogicUi.inventoryCategories.value.filter(cate => cate === val).length <= 1 || ''"
                                virtual-event="true"
                                modelvalue="Food"
          >
          </g-text-field-bs-stub>
          <div class="category-item__btn">
            <g-icon-stub>
              icon-delete2
            </g-icon-stub>
          </div>
        </div>
        <div class="category-item">
          <g-text-field-bs-stub rules="val => _inventoryLogicUi.inventoryCategories.value.filter(cate => cate === val).length <= 1 || ''"
                                virtual-event="true"
                                modelvalue="Fruit"
          >
          </g-text-field-bs-stub>
          <div class="category-item__btn">
            <g-icon-stub>
              icon-delete2
            </g-icon-stub>
          </div>
        </div>
        <div class="category-item">
          <g-text-field-bs-stub rules="val => _inventoryLogicUi.inventoryCategories.value.filter(cate => cate === val).length <= 1 || ''"
                                virtual-event="true"
                                modelvalue="Drink"
          >
          </g-text-field-bs-stub>
          <div class="category-item__btn">
            <g-icon-stub>
              icon-delete2
            </g-icon-stub>
          </div>
        </div>
      </div>
      <p>
        * inventory.onlyEmpty
      </p>
      <div class="dialog-action">
        <g-btn-bs-stub data-jest-addcategory="true"
                       icon="add"
                       background-color="#1271FF"
        >
          article.category
        </g-btn-bs-stub>
        <g-btn-bs-stub data-jest-complete="true"
                       background-color="#388E3C"
        >
          inventory.complete
        </g-btn-bs-stub>
      </div>
    </div>
    <div class="dialog-overlay">
    </div>
  </div>
</g-dialog-stub>
`;

exports[`Test inventory screen Case 2: Category dialog 2`] = `
<g-dialog-stub fullscreen="true"
               modelvalue="true"
               content-class="dialog-inventory-category"
>
  <div class="dialog">
    <div class="dialog-left">
      <div class="category">
        <div class="category-item">
          <g-text-field-bs-stub rules="val => _inventoryLogicUi.inventoryCategories.value.filter(cate => cate === val).length <= 1 || ''"
                                virtual-event="true"
                                modelvalue
          >
          </g-text-field-bs-stub>
          <div class="category-item__btn category-item__btn--delete">
            <g-icon-stub>
              icon-delete2
            </g-icon-stub>
          </div>
        </div>
        <div class="category-item">
          <g-text-field-bs-stub rules="val => _inventoryLogicUi.inventoryCategories.value.filter(cate => cate === val).length <= 1 || ''"
                                virtual-event="true"
                                modelvalue="Food"
          >
          </g-text-field-bs-stub>
          <div class="category-item__btn">
            <g-icon-stub>
              icon-delete2
            </g-icon-stub>
          </div>
        </div>
        <div class="category-item">
          <g-text-field-bs-stub rules="val => _inventoryLogicUi.inventoryCategories.value.filter(cate => cate === val).length <= 1 || ''"
                                virtual-event="true"
                                modelvalue="Fruit"
          >
          </g-text-field-bs-stub>
          <div class="category-item__btn">
            <g-icon-stub>
              icon-delete2
            </g-icon-stub>
          </div>
        </div>
        <div class="category-item">
          <g-text-field-bs-stub rules="val => _inventoryLogicUi.inventoryCategories.value.filter(cate => cate === val).length <= 1 || ''"
                                virtual-event="true"
                                modelvalue="Drink"
          >
          </g-text-field-bs-stub>
          <div class="category-item__btn">
            <g-icon-stub>
              icon-delete2
            </g-icon-stub>
          </div>
        </div>
      </div>
      <p>
        * inventory.onlyEmpty
      </p>
      <div class="dialog-action">
        <g-btn-bs-stub data-jest-addcategory="true"
                       icon="add"
                       background-color="#1271FF"
        >
          article.category
        </g-btn-bs-stub>
        <g-btn-bs-stub data-jest-complete="true"
                       background-color="#388E3C"
        >
          inventory.complete
        </g-btn-bs-stub>
      </div>
    </div>
    <div class="dialog-keyboard">
      <pos-keyboard-full-stub type="alpha-number">
      </pos-keyboard-full-stub>
    </div>
    <div class="dialog-overlay">
    </div>
  </div>
</g-dialog-stub>
`;

exports[`Test inventory screen Case 2: Category dialog 3`] = `
Array [
  Object {
    "_id": "0000000000000000000007cf",
    "available": false,
    "name": "Food",
  },
  Object {
    "_id": "0000000000000000000007d0",
    "available": false,
    "name": "Fruit",
  },
  Object {
    "_id": "0000000000000000000007d1",
    "available": false,
    "name": "Drink",
  },
]
`;
